x/stake/client/rest/tx.go

```diff
@@ -86,182 +74,92 @@ func delegationsRequestHandlerFn(cdc *codec.Codec, kb keys.Keybase, cliCtx conte
-			messages[i] = stake.MsgDelegate{
-				DelegatorAddr: delAddr,
-				ValidatorAddr: valAddr,
-				Delegation:    msg.Delegation,
-			}
-
-			i++
+		if !bytes.Equal(info.GetPubKey().Address(), req.DelegatorAddr) {
+			utils.WriteErrorResponse(w, http.StatusUnauthorized, "Must use own delegator address")
+			return
 		}
 
-		for _, msg := range req.BeginRedelegates {
-			delAddr, err := sdk.AccAddressFromBech32(msg.DelegatorAddr)
-			if err != nil {
-				utils.WriteErrorResponse(w, http.StatusInternalServerError, err.Error())
-				return
-			}
-
-			if !bytes.Equal(info.GetPubKey().Address(), delAddr) {
-				utils.WriteErrorResponse(w, http.StatusUnauthorized, "Must use own delegator address")
-				return
-			}
-
-			valSrcAddr, err := sdk.ValAddressFromBech32(msg.ValidatorSrcAddr)
-			if err != nil {
-				utils.WriteErrorResponse(w, http.StatusInternalServerError, err.Error())
-				return
-			}
-			valDstAddr, err := sdk.ValAddressFromBech32(msg.ValidatorDstAddr)
-			if err != nil {
-				utils.WriteErrorResponse(w, http.StatusInternalServerError, err.Error())
-				return
-			}
-
-			shares, err := sdk.NewDecFromStr(msg.SharesAmount)
-			if err != nil {
-				utils.WriteErrorResponse(w, http.StatusInternalServerError, err.Error())
-				return
-			}
-
-			messages[i] = stake.MsgBeginRedelegate{
-				DelegatorAddr:    delAddr,
-				ValidatorSrcAddr: valSrcAddr,
-				ValidatorDstAddr: valDstAddr,
-				SharesAmount:     shares,
-			}
-
-			i++
+		msg := stake.NewMsgDelegate(req.DelegatorAddr, req.ValidatorAddr, req.Delegation)
+		err = msg.ValidateBasic()
+		if err != nil {
+			utils.WriteErrorResponse(w, http.StatusBadRequest, err.Error())
+			return
 		}
 
-		for _, msg := range req.BeginUnbondings {
-			delAddr, err := sdk.AccAddressFromBech32(msg.DelegatorAddr)
-			if err != nil {
-				utils.WriteErrorResponse(w, http.StatusInternalServerError, err.Error())
-				return
-			}
-
-			if !bytes.Equal(info.GetPubKey().Address(), delAddr) {
-				utils.WriteErrorResponse(w, http.StatusUnauthorized, "Must use own delegator address")
-				return
-			}
-
-			valAddr, err := sdk.ValAddressFromBech32(msg.ValidatorAddr)
-			if err != nil {
-				utils.WriteErrorResponse(w, http.StatusInternalServerError, err.Error())
-				return
-			}
-
-			shares, err := sdk.NewDecFromStr(msg.SharesAmount)
-			if err != nil {
-				utils.WriteErrorResponse(w, http.StatusInternalServerError, err.Error())
-				return
-			}
-
-			messages[i] = stake.MsgBeginUnbonding{
-				DelegatorAddr: delAddr,
-				ValidatorAddr: valAddr,
-				SharesAmount:  shares,
-			}
+		utils.CompleteAndBroadcastTxREST(w, r, cliCtx, baseReq, []sdk.Msg{msg}, cdc)
+	}
+}
 
-			i++
-		}
+func postRedelegationsHandlerFn(cdc *codec.Codec, kb keys.Keybase, cliCtx context.CLIContext) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		var req msgBeginRedelegateInput
 
-		simulateGas, gas, err := client.ReadGasFlag(baseReq.Gas)
+		cliCtx, err := utils.ReadRESTReq(w, r, cdc, cliCtx, &req)
```

Comment:
Ughhh, CLIContext has really evolved. I think we should consider renaming it. Maybe `ClientContext`?

Reply:
It is! I just think it's more appropriate to call it a `ClientContext` at this point (e.g. look how it's being used here).